File: examples/schaefer/schaefer.R
hake <- read.table("schaefer.dat", header=TRUE)
names(hake) <- c("t", "C", "I")
parameters <- list(logR=-1.1, logK=8.0, logQ=-7.9, logSigma=-2.3)

require(TMB)
compile("schaefer.cpp")
dyn.load(dynlib("schaefer"))

################################################################################

model <- MakeADFun(hake, parameters)
fit <- nlminb(model$par, model$fn, model$gr)
rep <- sdreport(model)

print(summary(rep))

################################################################################

hake$B <- model$report()$B
hake$Ihat <- model$report()$Ihat

par(mfrow=c(1,2))
matplot(hake$t, hake[c("C","B")], type="l",
        xlab="Year", ylab="Biomass and Catch (kt)")
plot(I~t, hake, ylim=c(0,1.1*max(hake$I)), yaxs="i")
lines(Ihat~t, hake)

File: examples/schaefer/schaefer.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(C);
  DATA_VECTOR(I);
  int n = C.size();

  PARAMETER(logR);
  PARAMETER(logK);
  PARAMETER(logQ);
  PARAMETER(logSigma);
  Type r = exp(logR);
  Type k = exp(logK);
  Type q = exp(logQ);
  Type sigma = exp(logSigma);
  vector<Type> B(n);
  vector<Type> Ihat(n);
  Type f;

  B(0) = k;
  for(int t=0; t<(n-1); t++)
  {
    B(t+1) = abs(B(t) + r*B(t)*(1-B(t)/k) - C(t));
  }
  Ihat = q*B;
  f = -sum(dnorm(log(I), log(Ihat), sigma, true));

  ADREPORT(log(B)); // uncertainty
  REPORT(B);        // plot
  REPORT(Ihat);     // plot

  return f;
}
